File: src-tauri/src/models/assistant.rs
use serde::{Deserialize, Serialize};

// ── Sprite Types ─────────────────────────────────────────────────

/// Sprite metadata returned by list_sprites (filesystem scan)
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SpriteInfo {
    pub filename: String,
    pub display_name: String,
    pub source: SpriteSource,
    pub file_size_bytes: u64,
    pub created_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum SpriteSource {
    Prebuilt,
    Generated,
    Uploaded,
}

/// Per-cell crop offsets stored in JSON sidecar
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SpriteCropOffsets {
    #[serde(default = "SpriteCropOffsets::default_version")]
    pub version: u32,
    pub cells: Vec<CellOffset>,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
pub struct CellOffset {
    pub x: i32,
    pub y: i32,
}

impl SpriteCropOffsets {
    fn default_version() -> u32 {
        1
    }
}

impl Default for SpriteCropOffsets {
    fn default() -> Self {
        Self {
            version: 1,
            cells: vec![CellOffset { x: 0, y: 0 }; 8],
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AiPersonality {
    pub id: String,
    pub name: String,
    pub prompt_text: String,
    pub is_builtin: bool,
    pub created_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AiAvatar {
    pub id: String,
    pub name: String,
    pub personality_id: String,
    pub image_path: Option<String>,
    pub companion_role_id: Option<String>,
    pub companion_role_custom: Option<String>,
    pub is_active: bool,
    pub created_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CompanionRolePreset {
    pub id: String,
    pub name: String,
    pub description: String,
    pub system_prompt_text: String,
    #[serde(default)]
    pub is_builtin: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AiConversation {
    pub id: String,
    pub avatar_id: String,
    pub started_at: String,
    pub ended_at: Option<String>,
    pub summary: Option<String>,
    pub message_count: u32,
    /// Compaction status: 0 = pending, 1 = success, 2 = failed
    pub compacted: u8,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AiMessage {
    pub id: String,
    pub conversation_id: String,
    pub role: String,
    pub content: String,
    pub created_at: String,
    pub token_estimate: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AiMemory {
    pub id: String,
    pub avatar_id: String,
    pub conversation_id: Option<String>,
    pub content: String,
    pub importance: u32,
    pub category: String,
    pub is_system: bool,
    pub created_at: String,
    pub last_referenced: Option<String>,
    pub superseded_by: Option<String>,
    pub active: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AiDailyLog {
    pub id: String,
    pub avatar_id: String,
    pub conversation_id: String,
    pub log_date: String,
    pub summary: String,
    pub created_at: String,
}

/// Structured output from the compaction prompt (internal, not sent to frontend)
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CompactionResult {
    pub summary: String,
    /// Optional journal entry — a more personal/narrative account of the conversation.
    /// Falls back to `summary` if the LLM doesn't provide it.
    #[serde(default)]
    pub journal_entry: Option<String>,
    pub memories: Vec<CompactionMemory>,
    pub superseded_memories: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CompactionMemory {
    pub content: String,
    pub importance: u32,
    pub category: String,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_ai_memory_serde_camel_case() {
        let mem = AiMemory {
            id: "m1".into(),
            avatar_id: "a1".into(),
            conversation_id: Some("c1".into()),
            content: "test content".into(),
            importance: 5,
            category: "general".into(),
            is_system: false,
            created_at: "2026-02-27".into(),
            last_referenced: Some("2026-02-27".into()),
            superseded_by: None,
            active: true,
        };
        let json = serde_json::to_string(&mem).unwrap();
        assert!(json.contains("avatarId"));
        assert!(json.contains("conversationId"));
        assert!(json.contains("isSystem"));
        assert!(json.contains("lastReferenced"));
        assert!(json.contains("supersededBy"));
        assert!(!json.contains("avatar_id"));
        let back: AiMemory = serde_json::from_str(&json).unwrap();
        assert_eq!(back.id, "m1");
        assert_eq!(back.importance, 5);
    }

    #[test]
    fn test_ai_avatar_serde_camel_case() {
        let avatar = AiAvatar {
            id: "a1".into(),
            name: "TestBot".into(),
            personality_id: "p1".into(),
            image_path: None,
            companion_role_id: Some("gaming-companion".into()),
            companion_role_custom: None,
            is_active: true,
            created_at: "2026-02-27".into(),
        };
        let json = serde_json::to_string(&avatar).unwrap();
        assert!(json.contains("personalityId"));
        assert!(json.contains("isActive"));
        assert!(json.contains("imagePath"));
        assert!(json.contains("companionRoleId"));
        assert!(json.contains("companionRoleCustom"));
        assert!(!json.contains("personality_id"));
        assert!(!json.contains("companion_role_id"));
        let back: AiAvatar = serde_json::from_str(&json).unwrap();
        assert_eq!(back.name, "TestBot");
        assert_eq!(back.companion_role_id.as_deref(), Some("gaming-companion"));
        assert!(back.companion_role_custom.is_none());
    }

    #[test]
    fn test_companion_role_preset_serde() {
        let role = super::CompanionRolePreset {
            id: "gaming-companion".into(),
            name: "Gaming Companion".into(),
            description: "Balanced approach".into(),
            system_prompt_text: "a gaming companion".into(),
            is_builtin: true,
        };
        let json = serde_json::to_string(&role).unwrap();
        assert!(json.contains("systemPromptText"));
        assert!(!json.contains("system_prompt_text"));
        let back: super::CompanionRolePreset = serde_json::from_str(&json).unwrap();
        assert_eq!(back.id, "gaming-companion");
    }

    #[test]
    fn test_ai_personality_serde_camel_case() {
        let p = AiPersonality {
            id: "p1".into(),
            name: "Test".into(),
            prompt_text: "You are a test".into(),
            is_builtin: false,
            created_at: "2026-02-27".into(),
        };
        let json = serde_json::to_string(&p).unwrap();
        assert!(json.contains("promptText"));
        assert!(json.contains("isBuiltin"));
        assert!(!json.contains("prompt_text"));
        let back: AiPersonality = serde_json::from_str(&json).unwrap();
        assert_eq!(back.prompt_text, "You are a test");
    }

    #[test]
    fn test_ai_daily_log_serde_camel_case() {
        let log = AiDailyLog {
            id: "l1".into(),
            avatar_id: "a1".into(),
            conversation_id: "c1".into(),
            log_date: "2026-02-27".into(),
            summary: "Had a chat".into(),
            created_at: "2026-02-27".into(),
        };
        let json = serde_json::to_string(&log).unwrap();
        assert!(json.contains("avatarId"));
        assert!(json.contains("logDate"));
        assert!(!json.contains("avatar_id"));
        let back: AiDailyLog = serde_json::from_str(&json).unwrap();
        assert_eq!(back.log_date, "2026-02-27");
    }
}

File: src/types/assistant.ts
export interface AiPersonality {
  id: string;
  name: string;
  promptText: string;
  isBuiltin: boolean;
  createdAt: string;
}

export interface AiAvatar {
  id: string;
  name: string;
  personalityId: string;
  imagePath: string | null;
  companionRoleId: string | null;
  companionRoleCustom: string | null;
  isActive: boolean;
  createdAt: string;
  crossAvatarMemoryAccess: boolean;
  crossAvatarMemoryPrivate: boolean;
}

export interface AvatarStats {
  memoryCount: number;
  journalCount: number;
  createdAt: string;
}

export interface CompanionRolePreset {
  id: string;
  name: string;
  description: string;
  systemPromptText: string;
  isBuiltin: boolean;
}

export interface AiConversation {
  id: string;
  avatarId: string;
  startedAt: string;
  endedAt: string | null;
  summary: string | null;
  messageCount: number;
  /** Compaction status: 0 = pending, 1 = success, 2 = failed */
  compacted: number;
}

/** An image attached to a chat message. */
export interface ImageAttachment {
  mimeType: string;
  /** Base64-encoded image data. */
  data: string;
  /** Short caption generated after sending (for smart placeholders in older messages). */
  caption?: string;
}

/** Result of preparing an image for chat attachment. */
export interface PreparedImage {
  mimeType: string;
  data: string;
  previewUrl: string;
}

export interface AiMessage {
  id: string;
  conversationId: string;
  role: "user" | "assistant" | "system";
  content: string;
  createdAt: string;
  tokenEstimate: number;
  /** Encrypted JSON of ImageAttachment[] — decrypted on the Rust side for display. */
  attachments?: ImageAttachment[];
}

export interface AiMemory {
  id: string;
  avatarId: string;
  conversationId: string | null;
  content: string;
  importance: number;
  category: "preference" | "opinion" | "fact" | "general" | "system";
  isSystem: boolean;
  createdAt: string;
  lastReferenced: string | null;
  supersededBy: string | null;
  active: boolean;
}

export interface AiDailyLog {
  id: string;
  avatarId: string;
  conversationId: string;
  logDate: string;
  summary: string;
  createdAt: string;
}

export interface ConversationEndedPayload {
  conversationId: string;
  reason: "manual" | "timer";
}

/** Payload for `ai-user-message` cross-window sync event. */
export interface UserMessagePayload {
  conversationId: string;
  content: string;
  /** Unix timestamp (seconds) — used to deduplicate messages from the local window. */
  timestamp: number;
}

/** Payload for `ai-compaction-started` and `ai-compaction-complete` events. */
export interface CompactionEventPayload {
  conversationId: string;
}

export interface ActionSuggestion {
  actionId: string;
  description: string;
  tier: 1 | 2;
  requiresConfirmation: boolean;
}

export interface ReviewConfirmationData {
  gameId: string;
  gameName: string;
  stars: number;
  reviewText: string;
}

/** A resolved action returned from Rust validation + game name resolution. */
export interface ResolvedAction {
  /** The resolved action ID (game name replaced with UUID for game-targeting actions). */
  actionId: string;
  /** The original action ID as the AI generated it (e.g., "favorite:Elden Ring"). */
  originalActionId: string;
  /** Execution tier: 1 = auto-execute, 2 = confirmation required. */
  tier: number;
  /** Human-readable description for Tier 2 confirmation cards. */
  description?: string;
  /** Extra data (review text, star rating, note text, shelf name). */
  payload?: Record<string, unknown>;
  /** The resolved display name (e.g., "Elden Ring") for confirmation cards. */
  resolvedName?: string;
}

/** Result set from validate_and_resolve_ai_actions IPC. */
export interface ResolvedActionSet {
  actions: ResolvedAction[];
  rejectedCount: number;
}

/** Pipeline execution status for the action state machine. */
export type PipelineStatus =
  | "idle"
  | "running"
  | "paused"
  | "completed"
  | "canceled"
  | "delaying";

/** Result of executing a single action in the pipeline. */
export interface ActionResult {
  actionId: string;
  originalActionId: string;
  success: boolean;
  error?: string;
  executedAt: string;
  /** True when a Tier 2 action was explicitly confirmed by the user before execution. */
  confirmed?: boolean;
}

// ── Streaming Types ──────────────────────────────────────────────

/** Payload emitted via `ai-stream-chunk` for progressive text rendering. */
export interface StreamChunkPayload {
  conversationId: string;
  text: string;
}

// ── Sprite Types ─────────────────────────────────────────────────

export interface SpriteInfo {
  filename: string;
  displayName: string;
  source: "prebuilt" | "generated" | "uploaded";
  fileSizeBytes: number;
  createdAt: string;
}

export type Expression =
  | "neutral"
  | "speaking"
  | "listening"
  | "sleepy"
  | "happy"
  | "sad"
  | "interested"
  | "bored";

export const EXPRESSION_GRID: Expression[] = [
  "neutral",
  "speaking",
  "listening",
  "sleepy", // Row 0
  "happy",
  "sad",
  "interested",
  "bored", // Row 1
];

export const EXPRESSION_LABELS: Record<Expression, string> = {
  neutral: "Neutral",
  speaking: "Speaking",
  listening: "Listening",
  sleepy: "Sleepy",
  happy: "Happy",
  sad: "Sad",
  interested: "Interested",
  bored: "Bored",
};

export interface SpriteCropOffsets {
  version: number;
  cells: Array<{ x: number; y: number }>;
}
